test_runner/fixtures/neon_tenant.py

```diff
@@ -0,0 +1,352 @@
+
+        self.lock = threading.Lock()
+        self.envs: Dict[NeonEnvDiscriminants, NeonEnv] = {}
+
+        self.builders: List[NeonEnvBuilder] = []
+
+    def get_repo_dir(self, disc: NeonEnvDiscriminants) -> Path:
+        # FIXME use discriminants
+        randstr = "".join(
+            random.SystemRandom().choice(string.ascii_uppercase + string.digits) for _ in range(10)
+        )
+        s = Path(f"shared-{randstr}")
+        return self.top_output_dir / s
+
+    def get_or_create(self, disc: NeonEnvDiscriminants) -> NeonEnv:
+        with self.lock:
+            env = self.envs.get(disc)
+            if env is None:
+                builder = NeonEnvBuilder(
+                    repo_dir=self.get_repo_dir(disc),
+                    port_distributor=self.port_distributor,
+                    run_id=self.run_id,
+                    mock_s3_server=self.mock_s3_server,
+                    pg_distrib_dir=self.pg_distrib_dir,
+                    preserve_database_files=self.preserve_database_files,
+                    pg_version=PgVersion("17"),  # FIXME: this should go unused. Pass None?
+                    test_name="shared",  # FIXME
+                    test_output_dir=Path("shared"),  # FIXME
+                    test_overlay_dir=None,
+                    top_output_dir=self.top_output_dir,
+                    neon_binpath=disc.neon_binpath,
+                    pageserver_virtual_file_io_engine=disc.pageserver_virtual_file_io_engine,
+                    pageserver_aux_file_policy=disc.pageserver_aux_file_policy,
+                    pageserver_default_tenant_config_compaction_algorithm=disc.pageserver_default_tenant_config_compaction_algorithm,
+                    # FIXME: only support defaults for these currently
+                    # pageserver_remote_storage
+                    # pageserver_config_override
+                    # num_safekeepers
+                    # num_pageservers
+                    # safekeepers_id_start
+                    # safekeepers_enable_fsync
+                    # auth_enabled
+                    # rust_log_override
+                    # default_branch_name
+                    initial_tenant=None,  # FIXME should go unused
+                    initial_timeline=None,  # FIXME should go unused
+                    # safekeeper_extra_opts: Optional[list[str]] = None,
+                    # storage_controller_port_override: Optional[int] = None,
+                    # pageserver_io_buffer_alignment: Optional[int] = None,
+                )
+                env = builder.init_start()
+
+                self.envs[disc] = env
+        return env
+
+    def __enter__(self) -> "NeonSharedEnvs":
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc: Optional[BaseException],
+        tb: Optional[TracebackType],
+    ):
+        for env in self.envs.values():
+            env.stop(immediate=True)
```

Comment:
Do we need to pass in the immediate? Instead perhaps the test tenant fixture could, on teardown make sure if the tenant still exists (and needs to exist), it is completely flushed to remote storage (`http.timeline_checkpoint(..., waitUploaded=True)`).